File: apps/daemon/src/coagentia_daemon/buffer.py
"""遥测缓冲（契约 D §7 缓冲纪律 / §9.1 daemon/buffer/）：需 ack 类上报的离线落盘 + 重传。

三条独立缓冲：
- diagnostics.jsonl：重复可容忍（铁律 5），无客户端主键，ack 后按已发条数移除；
- usage.jsonl：以适配器 ULID 主键，exactly-once 去重根基；ack 后按 id 集合移除。
- check-finished.jsonl：以 run_id 自然键去重；server 落终态并 ack 后移除。

环形上限（constants.BUFFER_*）：溢出丢最旧并追加一条 daemon.buffer_overflow 诊断（丢弃计数可见）。
落盘跨 daemon 重启：每次变更以同目录临时文件原子重写。重传**不虚增**——ULID 落盘后不再生成。
"""

from __future__ import annotations

import contextlib
import json
import os
import tempfile
from pathlib import Path
from typing import Any

from coagentia_contracts.constants import (
    BUFFER_DIAGNOSTICS_MAX,
    BUFFER_USAGE_MAX,
)
from coagentia_contracts.daemon import (
    BufferedCounts,
    CheckFinishedData,
    DiagnosticEventIn,
    TokenUsageEventIn,
)

from coagentia_daemon.paths import DataPaths
from coagentia_daemon.util import now_iso

_OVERFLOW_TYPE = "daemon.buffer_overflow"


class TelemetryBuffer:
    """diagnostics / usage / check.finished 三缓冲 JSONL 落盘。"""

    def __init__(
        self,
        paths: DataPaths,
        *,
        diagnostics_max: int = BUFFER_DIAGNOSTICS_MAX,
        usage_max: int = BUFFER_USAGE_MAX,
    ) -> None:
        self._paths = paths
        self._diag_max = diagnostics_max
        self._usage_max = usage_max
        self._diag: list[dict[str, Any]] = []
        self._usage: list[dict[str, Any]] = []
        self._checks: list[dict[str, Any]] = []
        self._dropped_diag = 0
        self._dropped_usage = 0
        self._load()

    # ---------------------------------------------------------------- 落盘装载/重写

    @property
    def _diag_path(self) -> Path:
        return self._paths.buffer_dir / "diagnostics.jsonl"

    @property
    def _usage_path(self) -> Path:
        return self._paths.buffer_dir / "usage.jsonl"

    @property
    def _check_path(self) -> Path:
        return self._paths.buffer_dir / "check-finished.jsonl"

    def _load(self) -> None:
        self._paths.buffer_dir.mkdir(parents=True, exist_ok=True)
        self._diag = _read_jsonl(self._diag_path)
        self._usage = _read_jsonl(self._usage_path)
        self._checks = _read_jsonl(self._check_path)

    def _rewrite_diag(self) -> None:
        _write_jsonl(self._diag_path, self._diag)

    def _rewrite_usage(self) -> None:
        _write_jsonl(self._usage_path, self._usage)

    def _rewrite_checks(self) -> None:
        _write_jsonl(self._check_path, self._checks)

    # ---------------------------------------------------------------- 追加（含溢出处置）

    def append_diagnostic(self, event: DiagnosticEventIn) -> None:
        self._diag.append(event.model_dump(mode="json"))
        if len(self._diag) > self._diag_max:
            overflow = len(self._diag) - self._diag_max
            del self._diag[:overflow]
            self._dropped_diag += overflow
            self._append_overflow_marker("diagnostics", self._dropped_diag)
        self._rewrite_diag()

    def append_usage(self, event: TokenUsageEventIn) -> None:
        self._usage.append(event.model_dump(mode="json"))
        if len(self._usage) > self._usage_max:
            overflow = len(self._usage) - self._usage_max
            del self._usage[:overflow]
            self._dropped_usage += overflow
            # usage 溢出计入 diagnostics 缓冲（成本口径尽量不丢，但仍留痕）。
            self._append_overflow_marker("usage", self._dropped_usage)
        self._rewrite_usage()

    def append_check(self, event: CheckFinishedData) -> None:
        """check.finished 以 run_id 去重落盘；未 ack 前重启仍可原样重传。"""
        row = event.model_dump(mode="json")
        for index, current in enumerate(self._checks):
            if current.get("run_id") == event.run_id:
                self._checks[index] = row
                self._rewrite_checks()
                return
        self._checks.append(row)
        self._rewrite_checks()

    def _append_overflow_marker(self, buffer_name: str, dropped_total: int) -> None:
        """溢出留痕：追加一条 daemon.buffer_overflow 诊断（不再触发二次溢出判定）。"""
        marker = DiagnosticEventIn(
            type=_OVERFLOW_TYPE,
            payload={"buffer": buffer_name, "dropped_total": dropped_total},
            at=now_iso(),
        )
        self._diag.append(marker.model_dump(mode="json"))
        if len(self._diag) > self._diag_max:
            del self._diag[: len(self._diag) - self._diag_max]

    # ---------------------------------------------------------------- 读取/确认（重传语义）

    def peek_diagnostics(self, n: int) -> list[DiagnosticEventIn]:
        return [DiagnosticEventIn.model_validate(e) for e in self._diag[:n]]

    def peek_usage(self, n: int) -> list[TokenUsageEventIn]:
        return [TokenUsageEventIn.model_validate(e) for e in self._usage[:n]]

    def ack_diagnostics(self, count: int) -> None:
        """确认前 count 条已落库 → 移除（重复可容忍，按发送顺序移除）。"""
        if count <= 0:
            return
        del self._diag[:count]
        self._rewrite_diag()

    def ack_usage(self, ids: list[str]) -> None:
        """确认给定 ULID 已 exactly-once 落库 → 按 id 移除（未 ack 的保留待重传）。"""
        if not ids:
            return
        drop = set(ids)
        self._usage = [e for e in self._usage if e.get("id") not in drop]
        self._rewrite_usage()

    def peek_checks(self, n: int) -> list[CheckFinishedData]:
        return [CheckFinishedData.model_validate(e) for e in self._checks[:n]]

    def ack_checks(self, run_ids: list[str]) -> None:
        if not run_ids:
            return
        drop = set(run_ids)
        self._checks = [e for e in self._checks if e.get("run_id") not in drop]
        self._rewrite_checks()

    def find_check(self, run_id: str) -> CheckFinishedData | None:
        row = next((e for e in self._checks if e.get("run_id") == run_id), None)
        return CheckFinishedData.model_validate(row) if row is not None else None

    def counts(self) -> BufferedCounts:
        return BufferedCounts(diagnostics=len(self._diag), usage=len(self._usage))

    def has_diagnostics(self) -> bool:
        return bool(self._diag)

    def has_usage(self) -> bool:
        return bool(self._usage)

    def has_checks(self) -> bool:
        return bool(self._checks)


def _read_jsonl(path: Path) -> list[dict[str, Any]]:
    if not path.exists():
        return []
    out: list[dict[str, Any]] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            out.append(json.loads(line))
        except json.JSONDecodeError:
            continue
    return out


def _write_jsonl(path: Path, rows: list[dict[str, Any]]) -> None:
    """先持久化同目录临时文件，再原子替换正式 JSONL。"""
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, temp_name = tempfile.mkstemp(
        dir=path.parent,
        prefix=f".{path.name}.",
        suffix=".tmp",
    )
    temp_path = Path(temp_name)
    try:
        with os.fdopen(fd, "w", encoding="utf-8", newline="\n") as stream:
            for row in rows:
                stream.write(json.dumps(row, ensure_ascii=False) + "\n")
            stream.flush()
            os.fsync(stream.fileno())
        os.replace(temp_path, path)
    except BaseException:
        # fdopen 可能在进入 with 前失败；已关闭 fd 会安全命中 suppress。
        with contextlib.suppress(OSError):
            os.close(fd)
        with contextlib.suppress(OSError):
            temp_path.unlink()
        raise

File: apps/daemon-ts/src/buffer.ts
/**
 * 遥测缓冲（契约 D §7 缓冲纪律 / §9.1 daemon/buffer/）：需 ack 类上报的离线落盘 + 重传。
 * 对等基准 = apps/daemon buffer.py（零行为改进）。
 *
 * 五条独立缓冲：
 * - diagnostics.jsonl：重复可容忍（铁律 5），无客户端主键，ack 后按已发条数移除；
 * - usage.jsonl：以适配器 ULID 主键，exactly-once 去重根基；ack 后按 id 集合移除；
 * - check-finished.jsonl：以 run_id 自然键去重；server 落终态并 ack 后移除；
 * - deploy-log.jsonl：以 (deployment_id, chunk_seq) 去重；
 * - deploy-finished.jsonl：以 deployment_id 去重（已终态重发经 find 重报）。
 *
 * 环形上限（constants.BUFFER_*）：溢出丢最旧并追加一条 daemon.buffer_overflow 诊断（丢弃计数可见）。
 * 落盘跨 daemon 重启：每次变更以同目录临时文件**整文件原子重写**（temp 写入 + fsyncSync +
 * renameSync；win32 rename 可覆盖既有文件，对等 py mkstemp→fsync→os.replace）。
 * 重传**不虚增**——ULID 落盘后不再生成。O(n) 全量重写策略维持 py 原样（勿改追加制）。
 */

import { randomBytes } from 'node:crypto';
import * as fs from 'node:fs';
import * as path from 'node:path';

import type { BufferedCounts, CheckFinishedData, DeployFinishedData, DeployLogReportData, DiagnosticEventIn, TokenUsageEventIn } from '@coagentia/contracts-ts';

import { BUFFER_DIAGNOSTICS_MAX, BUFFER_USAGE_MAX } from './generated/constants.ts';
import type { DataPaths, JsonObject } from './paths.ts';
import { nowIso } from './util.ts';

const OVERFLOW_TYPE = 'daemon.buffer_overflow';

/**
 * 测试可替换的落盘操作接缝（对等 py 测试 monkeypatch `buffer_module.json.dumps` /
 * `buffer_module.os.fsync` / `buffer_module.os.replace`——node:fs 命名空间对象冻结不可
 * spy，故以本对象为唯一补丁点；生产路径行为与直接调用 fs 完全一致）。
 */
export const _io = {
  dumps: (row: JsonObject): string => JSON.stringify(row),
  fsync: (fd: number): void => {
    fs.fsyncSync(fd);
  },
  replace: (source: string, target: string): void => {
    fs.renameSync(source, target);
  },
};

/** diagnostics / usage / check.finished（+ deploy.log / deploy.finished）缓冲 JSONL 落盘。 */
export class TelemetryBuffer {
  private readonly _paths: DataPaths;
  private readonly _diagMax: number;
  private readonly _usageMax: number;
  private _diag: JsonObject[] = [];
  private _usage: JsonObject[] = [];
  private _checks: JsonObject[] = [];
  private _deployLogs: JsonObject[] = [];
  private _deployFinished: JsonObject[] = [];
  private _droppedDiag = 0;
  private _droppedUsage = 0;

  constructor(paths: DataPaths, opts: { diagnosticsMax?: number; usageMax?: number } = {}) {
    this._paths = paths;
    this._diagMax = opts.diagnosticsMax ?? BUFFER_DIAGNOSTICS_MAX;
    this._usageMax = opts.usageMax ?? BUFFER_USAGE_MAX;
    this._load();
  }

  // ---------------------------------------------------------------- 落盘装载/重写

  private get _diagPath(): string {
    return path.join(this._paths.bufferDir, 'diagnostics.jsonl');
  }

  private get _usagePath(): string {
    return path.join(this._paths.bufferDir, 'usage.jsonl');
  }

  private get _checkPath(): string {
    return path.join(this._paths.bufferDir, 'check-finished.jsonl');
  }

  private get _deployLogPath(): string {
    return path.join(this._paths.bufferDir, 'deploy-log.jsonl');
  }

  private get _deployFinishedPath(): string {
    return path.join(this._paths.bufferDir, 'deploy-finished.jsonl');
  }

  private _load(): void {
    fs.mkdirSync(this._paths.bufferDir, { recursive: true });
    this._diag = readJsonl(this._diagPath);
    this._usage = readJsonl(this._usagePath);
    this._checks = readJsonl(this._checkPath);
    this._deployLogs = readJsonl(this._deployLogPath);
    this._deployFinished = readJsonl(this._deployFinishedPath);
  }

  private _rewriteDiag(): void {
    writeJsonl(this._diagPath, this._diag);
  }

  private _rewriteUsage(): void {
    writeJsonl(this._usagePath, this._usage);
  }

  private _rewriteChecks(): void {
    writeJsonl(this._checkPath, this._checks);
  }

  private _rewriteDeployLogs(): void {
    writeJsonl(this._deployLogPath, this._deployLogs);
  }

  private _rewriteDeployFinished(): void {
    writeJsonl(this._deployFinishedPath, this._deployFinished);
  }

  // ---------------------------------------------------------------- 追加（含溢出处置）

  appendDiagnostic(event: DiagnosticEventIn): void {
    this._diag.push({ ...event });
    if (this._diag.length > this._diagMax) {
      const overflow = this._diag.length - this._diagMax;
      this._diag.splice(0, overflow);
      this._droppedDiag += overflow;
      this._appendOverflowMarker('diagnostics', this._droppedDiag);
    }
    this._rewriteDiag();
  }

  appendUsage(event: TokenUsageEventIn): void {
    this._usage.push({ ...event });
    if (this._usage.length > this._usageMax) {
      const overflow = this._usage.length - this._usageMax;
      this._usage.splice(0, overflow);
      this._droppedUsage += overflow;
      // usage 溢出计入 diagnostics 缓冲（成本口径尽量不丢，但仍留痕）。
      this._appendOverflowMarker('usage', this._droppedUsage);
    }
    this._rewriteUsage();
  }

  /** check.finished 以 run_id 去重落盘；未 ack 前重启仍可原样重传。 */
  appendCheck(event: CheckFinishedData): void {
    const row: JsonObject = { ...event };
    const index = this._checks.findIndex((current) => current['run_id'] === event.run_id);
    if (index >= 0) {
      this._checks[index] = row;
      this._rewriteChecks();
      return;
    }
    this._checks.push(row);
    this._rewriteChecks();
  }

  /** 溢出留痕：追加一条 daemon.buffer_overflow 诊断（不再触发二次溢出判定）。 */
  private _appendOverflowMarker(bufferName: string, droppedTotal: number): void {
    const marker: JsonObject = {
      type: OVERFLOW_TYPE,
      payload: { buffer: bufferName, dropped_total: droppedTotal },
      at: nowIso(),
    };
    this._diag.push(marker);
    if (this._diag.length > this._diagMax) {
      this._diag.splice(0, this._diag.length - this._diagMax);
    }
  }

  // ---------------------------------------------------------------- 读取/确认（重传语义）

  peekDiagnostics(n: number): DiagnosticEventIn[] {
    return this._diag.slice(0, n).map((e) => ({ ...e })) as unknown as DiagnosticEventIn[];
  }

  peekUsage(n: number): TokenUsageEventIn[] {
    return this._usage.slice(0, n).map((e) => ({ ...e })) as unknown as TokenUsageEventIn[];
  }

  /** 确认前 count 条已落库 → 移除（重复可容忍，按发送顺序移除）。 */
  ackDiagnostics(count: number): void {
    if (count <= 0) return;
    this._diag.splice(0, count);
    this._rewriteDiag();
  }

  /** 确认给定 ULID 已 exactly-once 落库 → 按 id 移除（未 ack 的保留待重传）。 */
  ackUsage(ids: string[]): void {
    if (ids.length === 0) return;
    const drop = new Set<unknown>(ids);
    this._usage = this._usage.filter((e) => !drop.has(e['id']));
    this._rewriteUsage();
  }

  peekChecks(n: number): CheckFinishedData[] {
    return this._checks.slice(0, n).map((e) => ({ ...e })) as unknown as CheckFinishedData[];
  }

  ackChecks(runIds: string[]): void {
    if (runIds.length === 0) return;
    const drop = new Set<unknown>(runIds);
    this._checks = this._checks.filter((e) => !drop.has(e['run_id']));
    this._rewriteChecks();
  }

  findCheck(runId: string): CheckFinishedData | null {
    const row = this._checks.find((e) => e['run_id'] === runId);
    return row !== undefined ? ({ ...row } as unknown as CheckFinishedData) : null;
  }

  // ---------------------------------------------------------------- deploy.log（去重键 =
  // (deployment_id, chunk_seq)；需 ack，server 按已收 max chunk_seq 去重）。

  /** deploy.log 以 (deployment_id, chunk_seq) 去重落盘；未 ack 前重启仍可原样重传。 */
  appendDeployLog(data: DeployLogReportData): void {
    const row: JsonObject = { ...data };
    const index = this._deployLogs.findIndex(
      (current) =>
        current['deployment_id'] === data.deployment_id && current['chunk_seq'] === data.chunk_seq,
    );
    if (index >= 0) {
      this._deployLogs[index] = row;
      this._rewriteDeployLogs();
      return;
    }
    this._deployLogs.push(row);
    this._rewriteDeployLogs();
  }

  peekDeployLogs(n: number): DeployLogReportData[] {
    return this._deployLogs.slice(0, n).map((e) => ({ ...e })) as unknown as DeployLogReportData[];
  }

  ackDeployLog(deploymentId: string, chunkSeq: number): void {
    this._deployLogs = this._deployLogs.filter(
      (e) => !(e['deployment_id'] === deploymentId && e['chunk_seq'] === chunkSeq),
    );
    this._rewriteDeployLogs();
  }

  hasDeployLogs(): boolean {
    return this._deployLogs.length > 0;
  }

  // ---------------------------------------------------------------- deploy.finished（去重键 =
  // deployment_id；需 ack，已终态重发经 find 重报）。

  appendDeployFinished(data: DeployFinishedData): void {
    const row: JsonObject = { ...data };
    const index = this._deployFinished.findIndex(
      (current) => current['deployment_id'] === data.deployment_id,
    );
    if (index >= 0) {
      this._deployFinished[index] = row;
      this._rewriteDeployFinished();
      return;
    }
    this._deployFinished.push(row);
    this._rewriteDeployFinished();
  }

  peekDeployFinished(n: number): DeployFinishedData[] {
    return this._deployFinished
      .slice(0, n)
      .map((e) => ({ ...e })) as unknown as DeployFinishedData[];
  }

  ackDeployFinished(deploymentIds: string[]): void {
    if (deploymentIds.length === 0) return;
    const drop = new Set<unknown>(deploymentIds);
    this._deployFinished = this._deployFinished.filter((e) => !drop.has(e['deployment_id']));
    this._rewriteDeployFinished();
  }

  hasDeployFinished(): boolean {
    return this._deployFinished.length > 0;
  }

  findDeployFinished(deploymentId: string): DeployFinishedData | null {
    const row = this._deployFinished.find((e) => e['deployment_id'] === deploymentId);
    return row !== undefined ? ({ ...row } as unknown as DeployFinishedData) : null;
  }

  counts(): BufferedCounts {
    return { diagnostics: this._diag.length, usage: this._usage.length };
  }

  hasDiagnostics(): boolean {
    return this._diag.length > 0;
  }

  hasUsage(): boolean {
    return this._usage.length > 0;
  }

  hasChecks(): boolean {
    return this._checks.length > 0;
  }
}

function readJsonl(filePath: string): JsonObject[] {
  if (!fs.existsSync(filePath)) return [];
  const out: JsonObject[] = [];
  for (const rawLine of fs.readFileSync(filePath, 'utf-8').split('\n')) {
    const line = rawLine.trim();
    if (!line) continue;
    try {
      out.push(JSON.parse(line) as JsonObject);
    } catch {
      continue; // 损坏行跳过（对等 py JSONDecodeError → continue）
    }
  }
  return out;
}

/** 先持久化同目录临时文件，再原子替换正式 JSONL。 */
function writeJsonl(filePath: string, rows: JsonObject[]): void {
  const parent = path.dirname(filePath);
  fs.mkdirSync(parent, { recursive: true });
  // 对等 py tempfile.mkstemp(dir=parent, prefix=f".{name}.", suffix=".tmp")：
  // 随机中缀 + 'wx' 独占创建（碰撞概率可忽略，不做 mkstemp 式重试）。
  const tempPath = path.join(
    parent,
    `.${path.basename(filePath)}.${randomBytes(6).toString('hex')}.tmp`,
  );
  const fd = fs.openSync(tempPath, 'wx');
  let fdOpen = true;
  try {
    for (const row of rows) {
      fs.writeSync(fd, `${_io.dumps(row)}\n`);
    }
    _io.fsync(fd);
    fs.closeSync(fd);
    fdOpen = false;
    _io.replace(tempPath, filePath);
  } catch (err) {
    // 写入/替换任何一步失败：关 fd、清临时文件、重抛（正式文件保持旧完整版）。
    if (fdOpen) {
      try {
        fs.closeSync(fd);
      } catch {
        // 已关闭等：对等 py contextlib.suppress(OSError)
      }
    }
    try {
      fs.unlinkSync(tempPath);
    } catch {
      // 不存在等：对等 py contextlib.suppress(OSError)
    }
    throw err;
  }
}
